### PR TITLE
Scripts for EBI molecular interaction ontology database

### PR DESCRIPTION
**Pull request by mengzhensun**

### CODE REVIEW

File: scripts/proteinInteractionEBI/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Scripts for importing ontology dataset from the European Bioinformatics Institute (EMBL-EBI)
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:1]: The BioOntologySchema.mcf should go into the schema directory I think.

File: scripts/proteinInteractionEBI/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+```
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:3]: nit: I would drop this hierarchy. Gets out of date when you add files, and the browsing github we get to see the same structure :)

File: scripts/proteinInteractionEBI/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# Scripts for importing ontology dataset from the European Bioinformatics Institute (EMBL-EBI)
+
+This directory stores all scripts used to import datasets from the European Bioinformatics Institute (EMBL-EBI). 
+Here we only import the three subsets of the ontologies: "interaction detection method", "interaction type" and "database citation", which are commonly used in protein-protein interactions. 
+
+## Usage
+
+To generate 'BioOntologySchemaEnum.mcf' which contains all the schemas and "psimi2dcid.txt" which contains paired PSI-MI identifiers and DCID, run:
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:8]: What is psimi2dcid.txt used for? (add to doc)

File: scripts/proteinInteractionEBI/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+```bash
+python3 parseEBI.py mi.owl 
+```
+
+If new reference sources which are not properties in dcs occur, 'BioEBINewSource.txt' containing such information will be generated as well.
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:14]: What should user do with BioEBINewSource.txt?

File: scripts/proteinInteractionEBI/README.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+![Tree Structure](./graph/ontologyTree.png)
+
+![Multiple Parent Node](./graph/multipleParent.png)
+
+The original data file (mi.owl) uses "is_a" and "part_of" as the relation property to connect the child node to the parent node, however we don't distinguish these two and a property "specializationOf" is used for the child-parent connection.
+
+![Original Data Sample](./graph/originalDataSample.png)
+
+We make each term as an enumeration node of three subtrees with root nodes as: "database citation", "interaction detection method", and "interaction type". There are two main concerns that we didn't import all the nodes. First, we focus on the ontologies in protein-protein interaction and these three categories are the most commonly used. Secondly, importing too many general terms may cause confusion in our Data Commons knowledge graph.
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:30]: I might reword as: "to avoid general terms from another ontology from polluting the Data Commons schema."

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    
+    return "\n".join(schemaPieceList), termDic['id'][0], dcid, newSourceMap
+
+
+def main(argv):
+    dbFile = argv[0]
+    with open(dbFile, 'r') as fp:
+        file = fp.read()
+    # clip exists in dcs already. Substitute with ClipInteraction
+    file = file.replace("name: clip\ndef", "name: clip interaction\ndef")
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:243]: Its good practice to be consistent with quotes throughout the file. Generally, I think prefer single to double.

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    '''
+    
+    id2node = {}
+    id2className = {}
+    # build nodes and create the id2node dictionary at first iteration
+    for termText in fileTerms:
+        if not termText.startswith("[Term]"):
+            continue
+        # idString example: "MI:0000"
+        idString = termText.split("\n")[1].split(" ")[1]
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:264]: Consider saving termText.split('\n') in a variable and using it (instead of splitting it multiple times).

File: scripts/proteinInteractionMINT/README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# Scripts for importing ontology dataset from the Molecular INTeraction database (MINT)
```
Comment (pradh) [inline scripts/proteinInteractionMINT/README.md:1]: Can we move MINT to another PR? This is large enough as such.

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    return "\n".join(schemaPieceList), termDic['id'][0], dcid, newSourceMap
+
+
+def main(argv):
+    dbFile = argv[0]
+    with open(dbFile, 'r') as fp:
+        file = fp.read()
+    # clip exists in dcs already. Substitute with ClipInteraction
+    file = file.replace("name: clip\ndef", "name: clip interaction\ndef")
+    fileTerms = file.split('\n\n')[1:]
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:244]: nit: Not sure what python style-guide you followed. But I think the internal google style guide suggests snake-case file_terms.

File: scripts/proteinInteractionEBI/README.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+Data is available for download from
+http://psidev.info/groups/controlled-vocabularies. It can also be browsed at https://www.ebi.ac.uk/ols/ontologies/mi. 
+The ontologies dictionary has a tree structure. Note here that one parent node can have multiple child nodes and one child node can have multiple parent nodes as well.
+
+![Tree Structure](./graph/ontologyTree.png)
+
+![Multiple Parent Node](./graph/multipleParent.png)
+
+The original data file (mi.owl) uses "is_a" and "part_of" as the relation property to connect the child node to the parent node, however we don't distinguish these two and a property "specializationOf" is used for the child-parent connection.
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:26]: To confirm, are you using part_of? I see you use "relationship".

File: scripts/proteinInteractionMINT/README.md (hunk 8 above)
Comment (mengzhensun) [reply scripts/proteinInteractionMINT/README.md:1]: Sure I will move MINT to another PR

File: scripts/proteinInteractionEBI/parseEBI.py (hunk 9 above)
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parseEBI.py:244]: Yea I found out this too. I will change the code style

File: scripts/proteinInteractionEBI/README.md (hunk 10 above)
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/README.md:26]: Yes both "part_of" and "is_a" are considered child -> parent linkage, which is the "specializationOf" property in dcs

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+        ''' 
+        if term.startswith("is_a"):
+            idStringList.append(term.split(" ")[1])
+        elif term.startswith("relationship"):
+            idStringList.append(term.split(" ")[2])
+        else: continue
+            
+    return idStringList
+
+class GetTreeValues(object):
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:105]: nit: This sounds more like a function name than a class name... TreeBuilder?

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+    def get_subset_id(self, node):
+        """Takes the string identifer such as "MI:1349" as the root node value
+            returns all the tree node values as a set."""
+        # reset the value set to be empty every function call
+        self.nodeValuesSet = set()
+        # run a DFS on the tree
+        self.dfs(node)
+        return self.nodeValuesSet
+
+    def dfs(self, node):
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:124]: This seems like a local function, so _dfs?

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+        idString = termList[1].split(" ")[1]
+        parentIdList = get_parent_id_list(termList[1:])
+        for pId in parentIdList:
+            id2node[pId].childList.append(id2node[idString])
+            id2node[idString].parentList.append(id2node[pId])
+
+    # get the idStrings for the three target set
+    dfsCaller = GetTreeValues()
+    interactionTypeIdSet = dfsCaller.get_subset_id(id2node["MI:0001"]) # root id: MI:0001 
+    detectionMethodIdSet = dfsCaller.get_subset_id(id2node["MI:0190"])# root id: MI:0190
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py]: Magic constants like this, would ideally use a literal name.

INTERACTION_TYPE_ROOT = 'MI:0001'

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+    interactionSourceIdset = dfsCaller.get_subset_id(id2node["MI:0444"]) # root id: MI:0444
+
+    # delete root node value from the set
+    interactionTypeIdSet.remove("MI:0001")
+    detectionMethodIdSet.remove("MI:0190")
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py]: If you're doing this for every call, can you just make this happen inside get_subset_id() ?

For instance, using an interface like this:

  dfsCaller = GetTreeValues(id2node)
  ... = dfsCaller.get_subset_id('MI:0001')
  ... = dfsCaller.get_subset_id('MI:0190')

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+        id: MI:0444 name: database citation
+    3. save the nodes in the three sets to the corresponding enumearation schema
+    '''
+    
+    id2node = {}
+    id2className = {}
+    # build nodes and create the id2node dictionary at first iteration
+    for termText in fileTerms:
+        if not termText.startswith("[Term]"):
+            continue
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:262]: Since in all the loops you only care about termText that begins with [Term], how about doing it once after splitting it in line 244.

  fileTerms = file.split('\n\n')[1:]
  fileTerms = [for t in fileTerms if termText.startswith("[Term]")]

File: scripts/proteinInteractionEBI/README.md [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+## Database format
+
+Data is available for download from
+http://psidev.info/groups/controlled-vocabularies. It can also be browsed at https://www.ebi.ac.uk/ols/ontologies/mi. 
+The ontologies dictionary has a tree structure. Note here that one parent node can have multiple child nodes and one child node can have multiple parent nodes as well.
+
+![Tree Structure](./graph/ontologyTree.png)
+
+![Multiple Parent Node](./graph/multipleParent.png)
```
Comment (pradh) [inline scripts/proteinInteractionEBI/README.md:24]: Is this something you created or downloaded from ebi? Does it work to link to the source instead of checking in?

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+
+    schemaList = []
+    psimi2dcid = []
+    newSourceMap = {'references':{}}
+    for termText in fileTerms:
+        if not termText.startswith('[Term]'):
+            continue
+        term = termText.split('\n')[1:]
+        schemaRes = get_schema_from_text(term,id2node,newSourceMap,id2className,
+                                      interactionTypeIdSet,detectionMethodIdSet,interactionSourceIdset)
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:303]: term, id2node, newSourceMap

If you can use pylint - http://google.github.io/styleguide/pyguide.html

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+    detectionMethodIdSet.remove('MI:0190')
+    interactionSourceIdset.remove('MI:0444')
+
+    setList = [interactionTypeIdSet, detectionMethodIdSet,interactionSourceIdset]
+    print('The schema amount of each Enum: interactionType, detectionMethod,interactionSource are ')
+    print([len(s) for s in setList])
+
+    schemaList = []
+    psimi2dcid = []
+    newSourceMap = {'references':{}}
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:297]: describe the map?

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+        writeList.append(sourceType)
+        for source in newSourceMap[sourceType]:
+            line = source + ': ' + newSourceMap[sourceType][source]
+            writeList.append(sourceType)
+        writeList.append('\n')
+    if writeList:
+        with open('BioEBINewSource.txt','w') as fp:
+            fp.write('\n'.join(writeList))
+
+
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:329]: Try to use absl library so you get FLAGS. That's a much nicer way to take arguments.

Comment (pradh) [reply scripts/proteinInteractionEBI/parseEBI.py:329]: Here is an example: https://github.com/datacommonsorg/data/blob/master/scripts/us_fed/treasury_constant_maturity_rates/generate_csv_and_mcf.py#L31-L41

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+    writeList = []
+    for sourceType in newSourceMap:
+        if not newSourceMap[sourceType]: continue
+        writeList.append(sourceType)
+        for source in newSourceMap[sourceType]:
+            line = source + ': ' + newSourceMap[sourceType][source]
+            writeList.append(sourceType)
+        writeList.append('\n')
+    if writeList:
+        with open('BioEBINewSource.txt','w') as fp:
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:326]: Its generally a good idea to take the file paths as flags: https://github.com/datacommonsorg/data/blob/master/scripts/us_fed/treasury_constant_maturity_rates/generate_csv_and_mcf.py#L40

That way you can also change on the command line.

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+        value = ': '.join(lineList[1:])
+        termDic[key].append(value)
+        
+    defLong = termDic['def'][0]
+    idStart = defLong.rfind('[')
+    description = defLong[1:idStart-1-2]
+    termDic['def'] = [description]
+    IDString = defLong[idStart+1:-1]
+    if len(IDString)>0:
+        IDList = defLong[idStart+1:-1].split(', ')
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:166]: Comment what is going on here?

File: scripts/proteinInteractionEBI/parseEBI.py [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+    Here we use this function for instance name. Thus it allows to start with a number
+    """
+    jointName = astring.title().replace(' ','')
+    # substitute except for  _, character, number
+    nonLegit = re.compile(r'[\W]+')
+    className = nonLegit.sub('', jointName)
+    return className
+
+def get_class_name(astring):
+    return get_class_name_helper(astring)
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parseEBI.py:22]: Why do you need two functions here?

File: scripts/proteinInteractionEBI/README.md (hunk 16 above)
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/README.md:24]: These are two screenshots showing the data format. I can put links here instead.

File: scripts/proteinInteractionEBI/parseEBI.py (hunk 22 above)
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parseEBI.py:22]: Sure I will delete the unnecessary functions.

File: scripts/proteinInteractionEBI/parse_ebi_test.py [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+    file_terms = file.split('\n\n')
+    file_terms = [term_text.split('\n') for term_text in file_terms
+                  if term_text.startswith('[Term]')]
+    return file_terms
+
+CONST_FILE_TERMS = get_file_terms(TEST_TEXT)
+
+CONST_INTERACTION_TYPE_ID_SET = set(['MI:0045', 'MI:0401'])
+
+class TestParseEbi(unittest.TestCase):
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parse_ebi_test.py:64]: These are tests are ok. But for future changes, you could simplify the amount of effort you put in testing and improve the end-to-end-ness of tests by structuring it as follows.

Make your core function, which you call from main(), like this:

   parse_ebi(input_file1, input_file2, ... , output_file1, output_file2)

Then you can prep sample input files based on a tiny subset of actual data, and prep expected output files.

Your test then will then do something like this:

def test_end_to_end(self):
  parse_ebi('testdata/ebi_input1.txt', 'testdata/ebi_input2.txt', ...., 'ebi_actual1.txt', 'ebi_actual2.txt')
  // assert that content of ebi_actual1.txt matches testdata/ebi_expected1.txt, and so on.

File: scripts/proteinInteractionEBI/parse_ebi_test.py [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+        schema_res = parse_ebi.get_schema_from_text(term, CONST_ID_TO_NODE,
+                                                    new_source_map, CONST_ID_TO_CLASS_NAME,
+                                                    CONST_INTERACTION_TYPE_ID_SET, set(), set())
+        self.assertEqual(schema_res[0],
+                         ('Node: dcid:ExperimentalInteractionDetection\n'
+                          'typeOf: dcs:InteractionTypeEnum\n'
+                          'name: "ExperimentalInteractionDetection"\npsimiID: "MI:0045"\n'
+                          'description: "Methods base"\npubMedID: "14755292"\n'
+                          'specializationOf: dcs:InteractionDetectionMethod\n'
+                          'descriptionUrl: "http://psidev.info/groups/controlled-vocabularies"'))
```
Comment (pradh) [inline scripts/proteinInteractionEBI/parse_ebi_test.py:108]: Make these as CONSTANTS enclosed in '''...'''?

File: scripts/proteinInteractionEBI/parse_ebi.py [hunk 25]
```diff
@@ -0,0 +1,367 @@
+ # Copyright 2020 Google LLC
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi.py:1]: Sorry for the inconvenience! I should have changed the file name at the last round- I will do that the next time. Currently you can see the changes according to your suggestions from the commits before d97654f.

File: scripts/proteinInteractionEBI/parse_ebi.py [hunk 26]
```diff
@@ -0,0 +1,367 @@
...
+    del argv
+    database_file = FLAGS.database_file
+    with open(database_file, 'r') as file_open:
+        file = file_open.read()
+    # clip exists in dcs already. Substitute with ClipInteraction
+    file = file.replace('name: clip\ndef', 'name: clip interaction\ndef')
+    file_terms = file.split('\n\n')[1:]
+    file_terms = [term_text.split('\n') for term_text in file_terms
+                  if term_text.startswith('[Term]')]
+#     Parsing Steps:
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi.py:308]: Fixed

File: scripts/proteinInteractionEBI/parse_ebi.py [hunk 27]
```diff
@@ -0,0 +1,367 @@
...
+    id_to_node = build_child_parent_link(file_terms, id_to_node)
+
+    # get the id_strings for the three target set
+    dfs_caller = TreeBuilder(id_to_node)
+
+    INTERACTION_TYPE_ROOT = 'MI:0001'
+    DETECTION_METHOD_ROOT = 'MI:0190'
+    INTERACTION_SOURCE_ROOT = 'MI:0444'
+
+    interaction_type_id_set = dfs_caller.get_subset_id(INTERACTION_TYPE_ROOT) # root id: MI:0001
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi.py:327]: Fixed

File: scripts/proteinInteractionEBI/parse_ebi.py [hunk 28]
```diff
@@ -0,0 +1,367 @@
...
+    for source_type in new_source_map:
+        if not new_source_map[source_type]:
+            continue
+        write_list.append(source_type)
+        for source in new_source_map[source_type]:
+            line = source + ': ' + new_source_map[source_type][source]
+            write_list.append(line)
+        write_list.append('\n')
+    if write_list:
+        with open('BioEBINewSource.txt', 'w') as file_open:
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi.py:272]: Fixed

File: scripts/proteinInteractionEBI/parse_ebi.py [hunk 29]
```diff
@@ -0,0 +1,367 @@
...
+    DETECTION_METHOD_ROOT = 'MI:0190'
+    INTERACTION_SOURCE_ROOT = 'MI:0444'
+
+    interaction_type_id_set = dfs_caller.get_subset_id(INTERACTION_TYPE_ROOT) # root id: MI:0001
+    detection_method_id_set = dfs_caller.get_subset_id(DETECTION_METHOD_ROOT)# root id: MI:0190
+    interaction_source_id_set = dfs_caller.get_subset_id(INTERACTION_SOURCE_ROOT) # root id: MI:0444
+
+    set_list = [interaction_type_id_set, detection_method_id_set, interaction_source_id_set]
+
+    schema_list = []
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi.py:333]: schemaMCF.mcf contains the schemas which were added into files in data/schema. I used to combine the schema mcf and data mcf into one file, and now I split them. So no need to read the schemaMCF file now :)

File: scripts/proteinInteractionEBI/parse_ebi_test.py [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+                  if term_text.startswith('[Term]')]
+    return file_terms
+
+CONST_FILE_TERMS = get_file_terms(TEST_TEXT)
+
+CONST_INTERACTION_TYPE_ID_SET = set(['MI:0045', 'MI:0401'])
+
+class TestParseEbi(unittest.TestCase):
+    """Test the functions in parse_ebi.py"""
+    def test_get_id_maps(self):
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi_test.py:66]: Fixed- pylint didn't show line of space. I will pay attention to this.

File: scripts/proteinInteractionEBI/parse_ebi_test.py (hunk 24 above)
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi_test.py]: Fixed.

File: scripts/proteinInteractionEBI/parse_ebi_test.py (hunk 23 above)
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/parse_ebi_test.py:89]: I see! I was testing the important functions separately here. The end-to-end way is definitely easier. Will do it next time :)

File: scripts/proteinInteractionEBI/README.md [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+python3 parse_ebi.py -f mi.owl 
+```
+
+To test the script, run:
+
+```bash
+python3 parse_ebi_test.py
+```
+
+If new reference sources which are not properties in dcs occur, 'BioEBINewSource.txt' containing such information will be generated as well. If this is the case, new properties should be created for the new source (similar to property "dcs:pubMedID"), and the script should handle the new source corespondingly.
```
[the remark this answers is not in the document]
Comment (mengzhensun) [reply scripts/proteinInteractionEBI/README.md]: Added more instructions.
